### PR TITLE
quosures in teal.goshawk

### LINKED ISSUE #146: Replace chunks with the quosure
Introduce Quosure in tmg.
- Please work on the feature branch related to this 26_new_chunks@main
- Assign to this issue https://github.com/insightsengineering/teal.goshawk/issues/152 and address both in the same PR as they need to be resolved in the same time (because merge_expression_module needs data)

-----

- [x] boxplot @nikolas-burkoff  including `constr_anl_chunks` needed for all modules
- [x] correlation plot @nikolas-burkoff 
- [x] density distribution plot @nikolas-burkoff 
- [x] lineplot @nikolas-burkoff 
- [x] scatterplot @nikolas-burkoff 
- [x] spaghetti plot @nikolas-burkoff

### PR DESCRIPTION
**Pull request by nikolas-burkoff**
Closes #146 and also closes https://github.com/insightsengineering/teal.goshawk/issues/152

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -224,8 +220,7 @@ constr_anl_chunks <- function(session, input, datasets, dataname, param_id, para
     validate(need(constraint_var, "select a constraint variable"))
 
     # note that filtered is false thus we cannot use anl_param()$ANL
-    ANL <- datasets$get_data(dataname, filtered = FALSE) # nolint
-
+    ANL <- data[[dataname]]()
```
Comment (nikolas-burkoff) [inline R/utils.R:223]: @gogonzo what do we do here?

File: R/tm_g_gh_correlationplot.R [hunk 2]
```diff
@@ -563,7 +558,7 @@ srv_g_correlationplot <- function(id,
       validate(need(constraint_var, "select a constraint variable"))
 
       # note that filtered is false thus we cannot use anl_param()$ANL
-      ANL <- datasets$get_data(dataname, filtered = FALSE) # nolint
+      ANL <- data[[dataname]]() # nolint
```
Comment (nikolas-burkoff) [inline R/tm_g_gh_correlationplot.R:561]: Also here this used to use filtered = FALSE

File: R/tm_g_gh_lineplot.R [hunk 3]
```diff
@@ -734,8 +744,8 @@ srv_lineplot <- function(id,
             hline_arb = .(hline_arb),
             hline_arb_label = .(hline_arb_label),
             hline_arb_color = .(hline_arb_color),
-            xtick = .(if (!methods::is(xtick, "waiver") && !is.null(xtick)) quote(xtick) else xtick),
-            xlabel = .(if (!methods::is(xtick, "waiver") && !is.null(xtick)) quote(xlabel) else xlabel),
+            xtick = .(if (!is.null(xtick)) quote(xtick) else xtick),
+            xlabel = .(if (!is.null(xtick)) quote(xlabel) else xlabel),
```
Comment (nikolas-burkoff) [inline R/tm_g_gh_lineplot.R:748]: Could not get what was here working hence the extra else if statement on line 710 - what happened was a list() without the waiver class got passed to goshawk which then complained 🤷‍♀️ 
image

File: R/tm_g_gh_spaghettiplot.R [hunk 4]
```diff
@@ -408,18 +410,47 @@ srv_g_spaghettiplot <- function(id,
...
-      teal.code::chunks_push(
-        chunks = private_chunks,
-        id = "g_spaghettiplot",
-        expression = bquote({
+
+      private_quosure <- anl_q()$quosure
+
+      # this code is needed to make sure the waiver attribute
+      # of ggplot2::waiver is correctly passed to goshawk's spaghettiplot
+      if (!methods::is(xtick, "waiver")) {
```
Comment (nikolas-burkoff) [inline R/tm_g_gh_spaghettiplot.R:426]: See https://github.com/insightsengineering/teal.goshawk/pull/185#discussion_r953748064

File: R/tm_g_gh_boxplot.R [hunk 5]
```diff
@@ -367,12 +368,13 @@ srv_g_boxplot <- function(id,
                           hline_vars_colors,
                           hline_vars_labels) {
   with_reporter <- !missing(reporter) && inherits(reporter, "Reporter")
+  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelApi")
```
Comment (mhallal1) [inline R/tm_g_gh_boxplot.R:371]:
```suggestion
  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelAPI")
```

Comment (mhallal1) [reply R/tm_g_gh_boxplot.R:371]: This applies to the other also

File: R/tm_g_gh_correlationplot.R [hunk 6]
```diff
@@ -435,9 +436,10 @@ srv_g_correlationplot <- function(id,
                                   vline_vars_colors,
                                   vline_vars_labels) {
   with_reporter <- !missing(reporter) && inherits(reporter, "Reporter")
+  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelApi")
```
Comment (nikolas-burkoff) [inline R/tm_g_gh_correlationplot.R:439]:
```suggestion
  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelAPI")
```

File: R/tm_g_gh_density_distribution_plot.R [hunk 7]
```diff
@@ -290,33 +291,34 @@ srv_g_density_distribution_plot <- function(id, # nolint
                                             plot_height,
                                             plot_width) {
   with_reporter <- !missing(reporter) && inherits(reporter, "Reporter")
+  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelApi")
```
Comment (nikolas-burkoff) [inline R/tm_g_gh_density_distribution_plot.R:294]:
```suggestion
  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelAPI")
```

File: R/tm_g_gh_lineplot.R [hunk 8]
```diff
@@ -351,8 +352,9 @@ srv_lineplot <- function(id,
                          plot_height,
                          plot_width) {
   with_reporter <- !missing(reporter) && inherits(reporter, "Reporter")
+  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelApi")
```
Comment (nikolas-burkoff) [inline R/tm_g_gh_lineplot.R:355]:
```suggestion
  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelAPI")
```

File: R/tm_g_gh_scatterplot.R [hunk 9]
```diff
@@ -276,26 +277,25 @@ srv_g_scatterplot <- function(id,
                               plot_height,
                               plot_width) {
   with_reporter <- !missing(reporter) && inherits(reporter, "Reporter")
+  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelApi")
```
Comment (nikolas-burkoff) [inline R/tm_g_gh_scatterplot.R:280]:
```suggestion
  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelAPI")
```

File: R/tm_g_gh_spaghettiplot.R [hunk 10]
```diff
@@ -374,24 +375,25 @@ srv_g_spaghettiplot <- function(id,
                                 hline_vars_colors,
                                 hline_vars_labels) {
   with_reporter <- !missing(reporter) && inherits(reporter, "Reporter")
+  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelApi")
```
Comment (nikolas-burkoff) [inline R/tm_g_gh_spaghettiplot.R:378]:
```suggestion
  with_filter <- !missing(filter_panel_api) && inherits(filter_panel_api, "FilterPanelAPI")
```

File: NEWS.md [hunk 11]
```diff
@@ -1,5 +1,10 @@
 # teal.goshawk 0.1.13.9013
 
+### Breking changes
+
+* Replaced `chunks` with simpler `Quosure` class.
+* Replaced `datasets` argument containing `FilteredData` with the new arguments `data` (list of reactive datasets) and `filter_panel_api` (`FilterPanelApi`).
```
Comment (Polkas) [inline NEWS.md:6]:
```suggestion
* Replaced `datasets` argument containing `FilteredData` with the new arguments `data` (list of reactive datasets) and `filter_panel_api` (`FilterPanelAPI`).
```

File: NEWS.md [hunk 12]
```diff
@@ -1,5 +1,10 @@
 # teal.goshawk 0.1.13.9013
 
+### Breking changes
```
Comment (mhallal1) [inline NEWS.md:3]:
```suggestion
### Breaking changes
```

**Review by mhallal1 — COMMENTED** [review]
I went through the code, checked for chunks presence and tested the apps -> LGTM

File: R/utils.R (hunk 1 above)
Comment (Polkas) [reply R/utils.R:223]: The ranges will be different now, it not mean wrong. They will follow the span under filtered data which still have sense.

https://github.com/insightsengineering/teal.goshawk/blob/0b4e14ce53e8bf60e8805d266218b15801fdf359/R/utils.R#L242-L269

File: R/tm_g_gh_correlationplot.R (hunk 2 above)
Comment (Polkas) [reply R/tm_g_gh_correlationplot.R:561]: The same as in other case we will get different result for `rng <- range(val, na.rm = TRUE)`, which still make sense.
We have to settle if we accept this change.

File: R/tm_g_gh_boxplot.R [hunk 13]
```diff
@@ -346,16 +346,17 @@ ui_g_boxplot <- function(id, ...) {
...
     ),
-    forms = get_rcode_ui(ns("rcode")),
+    forms = teal.widgets::verbatim_popup_ui(ns("rcode"), "Show R code"),
     pre_output = a$pre_output,
     post_output = a$post_output
   )
 }
 
 
 srv_g_boxplot <- function(id,
```
Comment (mhallal1) [inline R/tm_g_gh_boxplot.R:356]: Example app crashes when changing the `Data constraint` to `Screening` and setting the `SEX` filter then deselecting its components:
```
Warning: Error in observe: invalid contraint_varBASE2
```

File: R/tm_g_gh_density_distribution_plot.R [hunk 14]
```diff
@@ -272,15 +272,16 @@ ui_g_density_distribution_plot <- function(id, ...) {
...
       )
     ),
-    forms = get_rcode_ui(ns("rcode")),
+    forms = teal.widgets::verbatim_popup_ui(ns("rcode"), "Show R code"),
     pre_output = a$pre_output,
     post_output = a$post_output
   )
 }
 
 srv_g_density_distribution_plot <- function(id, # nolint
```
Comment (mhallal1) [inline R/tm_g_gh_density_distribution_plot.R:281]: Empty filter:
Screenshot from 2022-09-06 11-04-35

Comment (mhallal1) [reply R/tm_g_gh_density_distribution_plot.R:281]: Biomarker deselected:
Screenshot from 2022-09-06 11-05-46

Comment (mhallal1) [reply R/tm_g_gh_density_distribution_plot.R:281]: Treatment variable deselected:
Screenshot from 2022-09-06 11-06-16

File: R/tm_g_gh_correlationplot.R [hunk 15]
```diff
@@ -413,15 +413,16 @@ ui_g_correlationplot <- function(id, ...) {
...
       )
     ),
-    forms = get_rcode_ui(ns("rcode")),
+    forms = teal.widgets::verbatim_popup_ui(ns("rcode"), "Show R code"),
     pre_output = a$pre_output,
     post_output = a$post_output
   )
 }
 
 srv_g_correlationplot <- function(id,
```
Comment (mhallal1) [inline R/tm_g_gh_correlationplot.R:422]: Empty filter:
Screenshot from 2022-09-06 11-12-31

Comment (mhallal1) [reply R/tm_g_gh_correlationplot.R:422]: Biomarker deselected:
Screenshot from 2022-09-06 11-11-06

File: R/tm_g_gh_spaghettiplot.R [hunk 16]
```diff
@@ -356,8 +356,9 @@ g_ui_spaghettiplot <- function(id, ...) {
 
 
 srv_g_spaghettiplot <- function(id,
```
Comment (mhallal1) [inline R/tm_g_gh_spaghettiplot.R:358]: Default example app page:
Screenshot from 2022-09-06 11-20-19

Comment (gogonzo) [reply R/tm_g_gh_spaghettiplot.R:358]: https://github.com/insightsengineering/teal.code/issues/60

File: R/tm_g_gh_density_distribution_plot.R (hunk 14 above)
Comment (gogonzo) [reply R/tm_g_gh_density_distribution_plot.R:281]: I will try to fix these with `req()` calls

File: R/tm_g_gh_spaghettiplot.R (hunk 16 above)
Comment (gogonzo) [reply R/tm_g_gh_spaghettiplot.R:358]: I've fixed all red errors in teal.goshawk  by including `req()`

File: R/tm_g_gh_boxplot.R (hunk 13 above)
Comment (gogonzo) [reply R/tm_g_gh_boxplot.R:356]: fixed by adding validation in https://github.com/insightsengineering/teal.goshawk/pull/185/commits/09110658598c3f2dc159271da58270efaf700264

File: R/tm_g_gh_density_distribution_plot.R (hunk 14 above)
Comment (mhallal1) [reply R/tm_g_gh_density_distribution_plot.R:281]: with `Treatment variable deselected`, `Descriptive Statistics` still display a red error.

File: R/tm_g_gh_correlationplot.R (hunk 15 above)
Comment (mhallal1) [reply R/tm_g_gh_correlationplot.R:422]: empty filter still crashing the app.

File: R/tm_g_gh_spaghettiplot.R (hunk 16 above)
Comment (mhallal1) [reply R/tm_g_gh_spaghettiplot.R:358]: error still reproducible.

File: R/tm_g_gh_density_distribution_plot.R (hunk 14 above)
Comment (gogonzo) [reply R/tm_g_gh_density_distribution_plot.R:281]: fixed

File: R/tm_g_gh_spaghettiplot.R (hunk 16 above)
Comment (gogonzo) [reply R/tm_g_gh_spaghettiplot.R:358]: fixed

File: R/tm_g_gh_correlationplot.R (hunk 15 above)
Comment (gogonzo) [reply R/tm_g_gh_correlationplot.R:422]: I can't see the red message anymore

**Review by mhallal1 — APPROVED** [review]

**Pull request merged**